### PR TITLE
Modified behavior tree to allow immediate replanning upon tool path planner modification

### PR DESCRIPTION
**Pull request by gopikaajaykumar**
Edited the behavior tree and related to code to allow the user to re-generate the tool path plan immediately upon tool path planner modification

### CODE REVIEW

**Comment by marip8** [comment]
Can you add a screen capture of what this change does/looks like in the GUI?

**Comment by gopikaajaykumar** [comment]
image

When the user opens the tool path planner GUI to make modifications to the tool path plan after generating a tool path plan, the UI now has the "Generate Tool Path Plan" option enabled so they no longer have to navigate to the previous page to regenerate the tool path plan.

File: snp_application/include/snp_application/snp_widget.h [hunk 1]
```diff
@@ -12,9 +12,6 @@ namespace Ui
 class SNPWidget;
 }
 
-class QStackedWidget;
-class QTextEdit;
-
```
Comment (marip8) [inline snp_application/include/snp_application/snp_widget.h:17]: Don't remove; this was added in a recently merged PR

File: snp_application/include/snp_application/snp_widget.h [hunk 2]
```diff
@@ -26,8 +23,6 @@ class SNPWidget : public QWidget
   void runTreeWithThread();
 
   virtual BT::BehaviorTreeFactory createBTFactory(int ros_short_timeout, int ros_long_timeout);
-  QStackedWidget* getStackedWidget();
-  QTextEdit* getTextEdit();
```
Comment (marip8) [inline snp_application/include/snp_application/snp_widget.h:30]: Don't remove

File: snp_application/src/snp_widget.cpp [hunk 3]
```diff
@@ -242,14 +246,4 @@ void SNPWidget::runTreeWithThread()
...
-QStackedWidget* SNPWidget::getStackedWidget()
-{
-  return ui_->stacked_widget;
-}
-
-QTextEdit* SNPWidget::getTextEdit()
-{
-  return ui_->text_edit_log;
-}
-
```
Comment (marip8) [inline snp_application/src/snp_widget.cpp:254]: Don't remove

**Comment by marip8** [comment]
> Can you add a screen capture of what this change does/looks like in the GUI?

By this I meant that I would like for you to record a video of your screen while you demonstrate the changes. I think I understand what you are trying to do from the code changes, but I'm not entirely sure

**Comment by gopikaajaykumar** [comment]
Below is a video showing how the application works now (using snp_automate_2022). The relevant part showing tool path replanning starts at 00:20.

https://github.com/ros-industrial-consortium/scan_n_plan_workshop/assets/15222261/eef2892f-b7c1-47b8-b927-d0c70b6c0c27

**Comment by marip8** [comment]
I think there is too much "stateful" logic in this single block, and I think it would be better and more clearly implemented as a sub-tree or modification to the existing tree instead. If I understand this correctly, you are trying to use the TPP config tool button to interrupt the task of motion planning approval to put you back into the task of tool path planning approval. It seems like you could accomplish that with the existing blocks like this:

tree

Then you would only need to add the `tppconfig` button to the blackboard to get the behavior you are looking for

**Comment by gopikaajaykumar** [comment]
Thanks, I'll try the alternative approach you suggested

**Comment by marip8** [comment]
@gopikaajaykumar @DavidSpielman any follow up on whether the change to the BT file per my suggestion resulted in the behavior you were looking for?

**Comment by DavidSpielman** [comment]
> @gopikaajaykumar @DavidSpielman any follow up on whether the change to the BT file per my suggestion resulted in the behavior you were looking for?

I haven't gotten a chance to test this behavior yet.

**Comment by marip8** [comment]
Closing in favor of #119

**Comment by gopikaajaykumar** [comment]
@marip8 @DavidSpielman Sorry this is late, but I just tried the latest code with Michael's behavior tree update and (not sure if I am doing something wrong/misunderstanding) it seems like it's back to the original behavior where you have to click Back to get the Generate Tool Path Plan button to become enabled. 

[Screencast from 06-20-2024 05:32:05 PM.webm](https://github.com/ros-industrial-consortium/scan_n_plan_workshop/assets/15222261/582d34d2-b47b-4ae8-a218-d7af6d9c21bd)
^ Don't know if it's clear from the video but I click Back first to get the button to get enabled again after modifying the tool path plan parameter

**Pull request closed without merging**